### PR TITLE
Let InkWell/Ink/ancestor support GlobalKey so that splash does not stop when changing position.

### LINKED ISSUE #6751: When InkWell has a GlobalKey and changes position, splash should not stop

### PR DESCRIPTION
**Pull request by YeungKC**
## Description

Reproduction code: https://github.com/flutter/flutter/issues/6751#issuecomment-727621925

Previously `Element` `deactivate` or `activate` would stop the splash animation.

Now `Ink` and `InkWell` have `GlobalKey` or the ancestor widget has `GlobalKey` and both support splash animation without stopping when the widget moves (e.g. depth change).

Additional Scope of Changes.
  - `MergeableMaterial`
    -  `GlobalKey` position change so splash animation doesn't stop.

before:
image
after:
image

## Related Issues

Fixes #6751

Original #70607

## Tests

I added the following tests:

- ink_paint_test.dart 
  - When Ink has a GlobalKey and changes position, splash should not stop
- ink_well_test.dart
  - When InkWell/Ancestor has a GlobalKey and changes position, splash should not stop

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - Modify the test to refresh the GlobalKey with the same result.(https://github.com/flutter/flutter/pull/71138/files#diff-04e0901809bec10776d61f4d7d4d681fe0c101e445ca8ccee9fb9ceca6c551fcR1104)

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by YeungKC** [comment]
I accidentally closed the original PR, please review the latest code.

File: packages/flutter/lib/src/material/mergeable_material.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/mergeable_material.dart:597]: this is a very good change

**Comment by YeungKC** [comment]
Hi @Hixie, I got some advice from you earlier from @dkwingsmt  and now I've updated it, do you have time to look at it?

The main thing is to expose the active correlation in state.

**Comment by Hixie** [comment]
Consider the case where the parent isn't a RenderBox, but still does the debug intrinsics dance with its RenderBox children.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:1118]: what's the use case for this getter?

File: packages/flutter/lib/src/material/ink_well.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/ink_well.dart:1029]: You could also write this as something like:
```dart
   for (InkFeature splash in [ ...?splashes, ..._hightlights.values ])
     splash.active = active;
```
Not sure which is better, up to you.

File: packages/flutter/lib/src/material/ink_well.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/ink_well.dart:831]: is it possible that we'll be reactivated after this point?

File: packages/flutter/lib/src/material/ink_well.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/ink_well.dart:790]: would it make sense to just inline _dispose here? I only see one call to _dispose

File: packages/flutter/lib/src/material/ink_highlight.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/ink_highlight.dart:87]: why this rename?

Comment (Hixie) [reply packages/flutter/lib/src/material/ink_highlight.dart:87]: (i ask only because it is potentially breaking, so if there's not a compelling reason to do it, it might be worth not bothering)

File: packages/flutter/lib/src/material/ink_highlight.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/ink_highlight.dart:96]: same question here... seems like this will almost certainly affect someone, so if there's not a strong reason to do it...

File: packages/flutter/lib/src/material/material.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart:630]: maybe call this one "visible" to avoid the clash with the existing API?

File: packages/flutter/lib/src/material/material.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart:595]: this looks like it leaves `renderObject._inkFeatures` with a bunch of pointers to disposed objects, maybe instead we should call a method on the render object to clean itself up, and have that method do something like:

```dart
for (InkFeature feature in _inkFeatures)
  feature.dispose();
_inkFeatures = null;
```

File: packages/flutter/lib/src/widgets/framework.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:1338]: i think having this is reasonable.

Could we rename it "reactivate", to make it clear that it's not called when the State is first created?

File: packages/flutter/lib/src/widgets/framework.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:1329]: typo: reactivated

File: packages/flutter/lib/src/widgets/framework.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:1332]:
```suggestion
  /// If the [widget] or one of its ancestors has a [GlobalKey], it is possible for a State
  /// to be reactivated after being [deactivate]d, as part of being reparented in the tree.
```

**Comment by Hixie** [comment]
This PR is making huge progress, nice work. It's very close to being ready to land. Thank you so much for your contributions so far!

File: packages/flutter/lib/src/widgets/framework.dart
Comment (YeungKC) [reply packages/flutter/lib/src/widgets/framework.dart:1338]: I was thinking the same thing, but should `Element.activate` be rename as well?

File: packages/flutter/lib/src/material/material.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/material.dart:595]: I don't think so, because InkFeature has _RenderInkFeatures, which calls _RenderInkFeatures._removeFeature() when call InkFeature.dispose().

```dart
  void _removeFeature(InkFeature feature) {
    assert(_inkFeatures != null);
    _inkFeatures!.remove(feature);
    markNeedsPaint();
  }
```


I improved this part of the code and added comments to explain it.

File: packages/flutter/lib/src/material/ink_well.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/ink_well.dart:1029]: `...?_splashes`
This is awesome ！！！！！I didn't know this was possible before.

File: packages/flutter/lib/src/material/ink_well.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/ink_well.dart:843]: The error is caused by `highlightsExist` is true, which causes `wantKeepAlive` to be true.

This is caused by calling `Highlight.dispose` after `deactivate`.

Two ways have been found to reproduce this, call `Highlight.dispose` when `State.dispose`, if `_highlights.values` is more than one then `highlightsExist` is true.

The second is that `State.deactivate` is followed by callback to `_handleTapCancel`.


For the first one, we can set all `_highlights.values` to null and then execute `highlight.dispose` to solve the problem.

I don't have a good solution for the second one, so I'm using `State.active` to solve it.

And I think that exposing `State.active` will help to solve similar requirements or issues in the future.

**Comment by YeungKC** [comment]
@Hixie I've done some updates and parsing on the review comment, will you have time to review it again?

File: packages/flutter/lib/src/material/material.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/material.dart]: Or do you like this?

```dart
    if (_inkFeatures != null) {
      for (int i = _inkFeatures!.length - 1; i >= 0; i--){
        _inkFeatures![i].dispose();
      }
    }
```

I really like the use of `list?.isNotEmpty ? false`, it looks very intuitive and clean.

File: packages/flutter/lib/src/material/ink_well.dart
Comment (Hixie) [reply packages/flutter/lib/src/material/ink_well.dart:843]: Sorry I feel like I'm missing some critical information here! I was just wondering if we need to call updateKeepAlive in reactivate() since we don't call it when handleInkRemoval is called after deactivate. I'm not sure what error you're referring to in your response.

File: packages/flutter/lib/src/material/ink_well.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/ink_well.dart:1222]:
```suggestion
/// Animations triggered by an [InkWell] will survive their widget moving
/// due to [GlobalKey] reparenting, assuming that the nearest [Material]
/// ancestor is the same before and after the move.
```

**Comment by Hixie** [comment]
What happens if the GlobalKey move changes whether there is an ancestor Material, or changes it from one Material to another?

**Comment by YeungKC** [comment]
The splash is actually the material that controls the drawing, so if the material is changed, then the splash animation of the old material will stop, and the new material will not have the old splash reference, so no new splash will be generated.

File: packages/flutter/lib/src/material/ink_well.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/ink_well.dart:843]: Sorry, I misunderstood what you meant.

You are right, we should call `updateKeepAlive` in `reactivate`, because wantKeepAlive's value maybe changed.

But it should be better if we call `updateKeepAlive` in `AutomaticKeepAliveClientMixin. reactivate`?

Thanks for your guidance.

Comment (YeungKC) [reply packages/flutter/lib/src/material/ink_well.dart:843]: done, after reading the `AutomaticKeepAlive` code I realized that I shouldn't call `updateKeepAlive` in `AutomaticKeepAliveClientMixin.reactivate`.

**Comment by Hixie** [comment]
> The splash is actually the material that controls the paint, so if the material is changed, then the splash animation of the old material will stop, and the new material will not have the old splash reference, so no new splash will be generated.

Are there tests that verify this? I'm not quite sure I understand what you mean by "the splash is actually the material that controls the paint"... I don't see any logic in Ink's deactivate to check if the InkController is the same one as before, for example.

File: packages/flutter/test/material/ink_well_test.dart
Comment (Hixie) [inline packages/flutter/test/material/ink_well_test.dart:1351]: you need a pump without a duration here, before the pump with the duration, otherwise the duration will have no effect (because it'll just assume that the first frame after the tap just happened to jank hard).

same for other cases below

File: packages/flutter/test/material/ink_well_test.dart
Comment (Hixie) [inline packages/flutter/test/material/ink_well_test.dart:1388]: per our style guide, "onFoo" must be a closure. This should be called something like "tapChangesWrap" or "changeWrapOnTap" or something.

File: packages/flutter/test/material/ink_well_test.dart
Comment (Hixie) [inline packages/flutter/test/material/ink_well_test.dart:1367]: i don't understand why this test fails with line 1359 set to false and line 1367 set to true. What's the difference between those two cases?

Comment (Hixie) [reply packages/flutter/test/material/ink_well_test.dart:1367]: oh, i see, in the cases above the ink well is reparented so fast that the first frame has it moved already.

Comment (Hixie) [reply packages/flutter/test/material/ink_well_test.dart:1367]: i think the test would be more useful with checks immediately after tap (pump no duration), then after a short duration but before the move, then after the move, then at the end (after "settle" or the big pump).

File: packages/flutter/test/material/ink_well_test.dart
Comment (Hixie) [inline packages/flutter/test/material/ink_well_test.dart:1353]: when using pumpAndSettle you should check the return value.

Better to just pump with a big value.

File: packages/flutter/test/material/ink_well_test.dart
Comment (Hixie) [inline packages/flutter/test/material/ink_well_test.dart:1451]: these blank lines are mildly confusing. Blank lines read like paragraph breaks, but here you have one inside a nested closure and then outside that closure in the same statement, so it's not clear what is going on. I would just remove all the blank lines and let the indentation speak for itself.

File: packages/flutter/test/material/ink_well_test.dart
Comment (Hixie) [inline packages/flutter/test/material/ink_well_test.dart:1486]: fwiw this test is written in a very confusing way (calling a closure with a closure and so on). Try printing out the code and then drawing lines showing the code flow -- it jumps up and down and up and down a bunch of times.

It would be clearer if there were no closures, maybe by making a stateful widget that wraps the behaviour you're testing (what is currently in the closure you pass to `test`, except using State instead of a StatefulBuilder).

File: packages/flutter/test/material/ink_well_test.dart
Comment (Hixie) [inline packages/flutter/test/material/ink_well_test.dart:1534]: tests should be as simple as possible. Do you need the Scaffold here for example?

File: packages/flutter/test/material/ink_well_test.dart
Comment (dkwingsmt) [inline packages/flutter/test/material/ink_well_test.dart:1059]: This test name does not sound like the behavior after this PR. Maybe we should revamp the test into the current behavior.

**Review by dkwingsmt — CHANGES_REQUESTED** [review]
Mostly some documentation changes to the framework change.

File: packages/flutter/lib/src/material/material.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/material.dart:558]: This is an error, if _inkFeatures is null then it will return null, we can choose to use foo?.bar == true or foo?.bar ?? false.

File: packages/flutter/test/material/ink_well_test.dart
Comment (YeungKC) [reply packages/flutter/test/material/ink_well_test.dart]: As a result of our previous discussion, this was an outdated test, so I removed it.

**Comment by Piinks** [comment]
Hi @YeungKC would you like to continue with this change? It looks there there are merge conflicts currently that need to be resolved.

**Comment by dkwingsmt** [comment]
Oh I'm reviewing it. I'll let him know to resolve once its done.

**Review by dkwingsmt — CHANGES_REQUESTED** [review]
Full review: Looks great (especially the new tests) except for one question. Also there's a conflict.

**Comment by dkwingsmt** [comment]
@Hixie I think the PR looks good but would you like to give a final approval for the change on `State.reactivate`?

File: packages/flutter/lib/src/widgets/framework.dart
Comment (Hixie) [reply packages/flutter/lib/src/widgets/framework.dart:938]: Unless there's a good use case I'd rather we didn't add this. In general exposing lifecycles like this is not necessary and may encourage poor practices.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:1123]: actually it always calls it when removing the widget, not just when it has a global key.

File: packages/flutter/lib/src/material/ink_well.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/ink_well.dart:783]: If _splashes and _highlights are empty, then null is returned. Nothing needs to be done in this case.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (YeungKC) [reply packages/flutter/lib/src/widgets/framework.dart:1166]: Element.unmount will call State.dispose They also check that state._debugLifecycleState == _StateLifecycle.defunct, and I should probably do a duplicate check at State.reactivate as well But... does that really help improve code quality?

File: packages/flutter/lib/src/widgets/framework.dart
Comment (YeungKC) [reply packages/flutter/lib/src/widgets/framework.dart:939]: Maybe it has been too long and we have forgotten some details.

This is so that updateKeepAlive can be executed at the right time. If Material is calling dispose, but InkWell is inactive, then InkHighlight.remove is executed. If we don't know if it's active at this time, this won't work.

Before this I tried manually recording active inside inkwell.state, but I felt it was very ugly...

btw, sorry I ignored the reply.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (YeungKC) [reply packages/flutter/lib/src/widgets/framework.dart]: You're right, I got confused.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (dkwingsmt) [reply packages/flutter/lib/src/widgets/framework.dart:1165]: It's mostly State checking that it's own lifecycle methods are called in a regular way. Element might be refactored some time and these external guarantees might be accidentally removed or altered. If this kind of regularity is something the API guarantees, then it's always beneficial to have it asserted.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (dkwingsmt) [inline packages/flutter/lib/src/widgets/framework.dart:1148]:
```suggestion
      return _debugActive == false;
    }());
```

File: packages/flutter/lib/src/widgets/framework.dart
Comment (dkwingsmt) [inline packages/flutter/lib/src/widgets/framework.dart:1172]:
```suggestion
      _debugActive = !_debugActive;
      return _debugActive == true;
    }());
```

File: packages/flutter/lib/src/widgets/framework.dart
Comment (dkwingsmt) [inline packages/flutter/lib/src/widgets/framework.dart:934]:
```suggestion
  /// they are called alternatively.
  ///
  /// This field is not set in release mode.
```

Comment (YeungKC) [reply packages/flutter/lib/src/widgets/framework.dart:934]: done

**Review by dkwingsmt — COMMENTED** [review]
LGTM. I'm leaving the final word to Hixie.

**Review by dkwingsmt — APPROVED** [review]
LGTM. I'm leaving the final word to Hixie.

File: packages/flutter/lib/src/widgets/framework.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart:1171]: `== true` is redundant here. I'm guessing that's one of the things the analyzer shard is complaining about, though I haven't checked.

File: packages/flutter/lib/src/material/material.dart
Comment (Hixie) [inline packages/flutter/lib/src/material/material.dart:634]: The docs for this should be clearer about the responsibilities of the caller when this is changed. (For example, do you have to setState on something?)

**Comment by Hixie** [comment]
This is very close! Left some very minor comments but the bigger problem is it looks like tests are failing.

File: packages/flutter/lib/src/material/material.dart
Comment (YeungKC) [reply packages/flutter/lib/src/material/material.dart:640]: Updated.

**Comment by Hixie** [comment]
LGTM!

**Comment by dkwingsmt** [comment]
@YeungKC This PR is causing an internal breakage, which I've isolated as follows:
```dart
  testWidgets('a test', (WidgetTester tester) async {
      final ThemeData theme = ThemeData(
      // Colors.
        splashFactory: InkRipple.splashFactory,
      );

      await tester.pumpWidget(MaterialApp(
        theme: theme,
        home: Scaffold(
          body: InputChip(
            label: Text('Chip'),
            onPressed: () {},
          ),
        ),
      ));

      await tester.press(find.byType(RawChip));
      await tester.pump();
  });
```

Can you investigate before landing again?

**Comment by dkwingsmt** [comment]
Another breakage:
```dart


  testWidgets('test2', (WidgetTester tester) async {
      await tester.pumpWidget(MaterialApp(
        home: Scaffold(
          body: TestRetryTestApp()
        ),
      ));

      await tester.pump();

      await tester.tap(find.byType(IconButton));

      await expectLater(find.byType(TestRetryTestApp), matchesGoldenFile('mobile_app_tile_android'));
  });


class RetryWidget extends StatelessWidget {
  RetryWidget({required this.onClick, required this.label});

  final VoidCallback onClick;
  final String label;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: IconButton(
        icon: const Icon(Icons.ac_unit),
        onPressed: () => onClick(),
      ),
    );
  }
}

class TestRetryTestApp extends StatefulWidget {
  @override
  TestRetryTestAppState createState() => TestRetryTestAppState();
}
class TestRetryTestAppState extends State<TestRetryTestApp> {
  TestRetryTestAppState();

  bool loading = false;

  void _loadAppData() {
    loading = true;
  }

  @override
  Widget build(BuildContext context) {
    if (loading) {
      return Container();
    }
    return RetryWidget(
      onClick: _loadAppData,
      label: 'Loading',
    );
  }
}
```

**Pull request merged**
